### ISSUE TITLE
JAX Dirichlet Log PDF output different thatn SciPy

### ISSUE THREAD

**Issue by allentran** [google/jax#5711]
```
jax.scipy.stats.dirichlet.logpdf(np.array([9.2820191190831968786056904718861915171146e-01,
1.7047643862547270932061849180172430351377e-03,
3.8956840931979931121065252597190919914283e-03,
5.3582474241543359694261994263797532767057e-02,
6.7548574969396027109502789187445159768686e-03,
2.8701918527402454049679558778507271199487e-03,
2.9901160210044600180900875585621179197915e-03]), np.array([2.0922036170959472656250000000000000000000e+00,
2.5644928216934204101562500000000000000000e-01,
1.0264828205108642578125000000000000000000e+00,
6.1082868576049804687500000000000000000000e+00,
5.9878396987915039062500000000000000000000e-01,
2.6282947063446044921875000000000000000000e+00,
7.7161555290222167968750000000000000000000e+00], dtype=np.float32))
```
outputs `-inf` whereas 
```
scipy.stats.dirichlet.logpdf(np.array([9.2820191190831968786056904718861915171146e-01,
1.7047643862547270932061849180172430351377e-03,
3.8956840931979931121065252597190919914283e-03,
5.3582474241543359694261994263797532767057e-02,
6.7548574969396027109502789187445159768686e-03,
2.8701918527402454049679558778507271199487e-03,
2.9901160210044600180900875585621179197915e-03]), np.array([2.0922036170959472656250000000000000000000e+00,
2.5644928216934204101562500000000000000000e-01,
1.0264828205108642578125000000000000000000e+00,
6.1082868576049804687500000000000000000000e+00,
5.9878396987915039062500000000000000000000e-01,
2.6282947063446044921875000000000000000000e+00,
7.7161555290222167968750000000000000000000e+00], dtype=np.float32))
```
outputs -31.38.

I tried using 64-bit floats but that didn't work either.

**Comment by jakevdp** [comment]
What version of JAX are you using? On version 0.2.9 I get the correct output.

**Comment by allentran** [comment]
Same version. It might be a precision issue. I can repro with:
```
from jax.config import config
config.update("jax_enable_x64", True)
```
at the top of the script and cannot without those additional lines.

**Comment by allentran** [comment]
Digging around the JAX code, I think it is this (line 38, `dirichlet.py`):
```
    return jnp.where(_is_simplex(x), log_probs, -jnp.inf)
```
where
```
def _is_simplex(x):
    x_sum = jnp.sum(x, axis=-1)
    return jnp.all(x > 0, axis=-1) & (x_sum <= 1) & (x_sum > 1 - 1e-6)
```
Something about the precision makes the sum of those fractions marginally greater than 1.

**Comment by jakevdp** [comment]
Ah, I see. It looks like JAX and scipy use different criteria to check if the value is close enough to 1. Jax is here: https://github.com/google/jax/blob/ab21409e0b2a561f217c833ca6f915b9dcefffd0/jax/_src/scipy/stats/dirichlet.py#L27

And scipy is here: https://github.com/scipy/scipy/blob/fa119c98b18e3dfc6e158e22e806831f40d660d3/scipy/stats/_multivariate.py#L1287

Basically, jax defines the valid input range as `1 - 1E-6 < x_sum <= 1` and scipy defines it as `1 - 10E-10 <= x_sum <= 1 + 10E-10`

The fix here would be to make these constraints match.

**Comment by allentran** [comment]
Nice. 

The SciPy tolerance seems more forgiving, normalizing within the current JAX tolerance is trickier than it seems (dividing by the sum of the array isn't sufficient, so some form of truncating precision and then summing might have to do).

**Comment by jakevdp** [comment]
I started working on this, and it looks like the problems run even deeper. Due to a copy/paste typo, we never even test dirichlet distributions (note `cauchy` here): https://github.com/google/jax/blob/3c87a36831109ddcbc95e3844108c78f51d00300/tests/scipy_stats_test.py#L166-L170

And because of this we never discovered that the implementation in JAX is incompatible with the implementation in scipy, which requires 1D inputs!

**Comment by jakevdp** [comment]
For example:
```python
import numpy as np
import scipy.stats
import jax.scipy.stats

np.random.seed(0)
x = np.random.rand(4)
x /= x.sum()
alpha = np.ones(5)

print(scipy.stats.dirichlet.logpdf(x, alpha))
# 3.1780538303479458

print(jax.scipy.stats.dirichlet.logpdf(x, alpha))
# ValueError: Incompatible shapes for broadcasting: ((5,), (4,), (1,))
```

**Comment by jakevdp** [comment]
Thanks for the report! This revealed some far deeper bugs, which are now fixed by #5740

**Comment by pawel-czyz** [comment]
Hi, I wonder if I did something wrong or whether the bug reappeared in 0.2.21:

```
import jax
import numpy as np
import scipy.stats

for dtype in [np.float32, np.float64]:
    x = np.array([0.5, 0.5], dtype=dtype)
    alpha = np.ones(3, dtype=dtype)

    print(f"Data type: {dtype}")
    print(f"  Jax:   {jax.scipy.stats.dirichlet.logpdf(x, alpha)}")
    print(f"  SciPy: {scipy.stats.dirichlet.logpdf(x, alpha)}")
    print("\n")
```

gives:
```
Data type: <class 'numpy.float32'>
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
  Jax:   -inf
  SciPy: 0.6931471824645996


Data type: <class 'numpy.float64'>
  Jax:   -inf
  SciPy: 0.6931471805599453
```
I use:
```
python 3.9.7
jax    0.2.21
jaxlib 0.1.71
scipy 1.7.1
```
